pulp_smash/tests/rpm/api_v2/test_force_full.py

```diff
@@ -0,0 +1,200 @@
+    task_ids = set()
+    tasks_search = '/pulp/api/v2/tasks/search/'
+    search_criteria = {'filters': {'state': {'$ne': 'running'}}}
+    polls = 0
+    for task in tasks:
+        for spawned in task['spawned_tasks']:
+            task_ids.add(spawned['task_id'])
+    results = {}
+    while task_ids:
+        search_criteria['filters']['task_id'] = {'$in': list(task_ids)}
+        finished_tasks = client.post(urljoin(cfg.base_url, tasks_search),
+                                     {'criteria': search_criteria}).json()
+        for task in finished_tasks:
+            results[task['task_id']] = task
+            task_ids -= set([task['task_id']])
+        time.sleep(2)
+        polls += 1
+        if polls > 20:
+            raise WaitForTasksTimeoutError()
+    return results
+
+
+def _copy_units(server_config, dest_repo_href, source_repo_id, units):
+    """Copy units ``units`` from the repository ``repo1`` to ``repo2``.
+
+    Return the JSON-decoded response body.
+    """
+    path = urljoin(dest_repo_href, 'actions/associate/')
+    type_ids = list(set([unit['unit_type_id'] for unit in units]))
+    units_ids = [unit['unit_id'] for unit in units]
+    body = {'criteria': {'filters': {'unit': {'_id': {'$in': units_ids}}},
+                         'type_ids': type_ids},
+            'source_repo_id': source_repo_id}
+    return api.Client(server_config).post(path, body).json()
+
+
+class ForceFullTestCase(utils.BaseAPITestCase):
+    """Test Forcefull feature."""
+
+    @classmethod
+    def setup_repo(cls, client):
+        """Create repository and associate distributor with it."""
+        body = gen_repo()
+        body['importer_config']['feed'] = RPM_FEED_URL
+        repo = client.post(REPOSITORY_PATH, body).json()
+
+        # Add a distributor.
+        distributor = client.post(
+            urljoin(repo['_href'], 'distributors/'),
+            gen_distributor(),
+        ).json()
+        return (repo, distributor)
+
+    @classmethod
+    def setUpClass(cls):
+        """Create repository, associate, publish, associate, publish, publish.
+
+        Following steps are executed:
+
+        1. Create repository 1 with feed, sync and publish it.
+        2. Create repository 2 with feed
+        3. Copy 3 random units from repository 1 to repository 2
+        4. Publish repository 2
+        5. Copy 3 random units from repository 1 to repository 2
+        6. Publish repository 2
+        7. Publish repository 2 with force_full
+        """
+        super(ForceFullTestCase, cls).setUpClass()
+        cls.responses = {}
+        client = api.Client(cls.cfg)
+
+        # Create repo 1
+        repo1, _ = cls.setup_repo(client)
+        cls.resources.add(repo1['_href'])  # mark for deletion
+
+        # Sync repo 1
+        cls.responses['sync'] = utils.sync_repo(cls.cfg, repo1['_href'])
+        report = cls.responses['sync'].json()
+        next(api.poll_spawned_tasks(cls.cfg, report))
```

Comment:
By default, the `api.Client` class treats all HTTP 202 responses from the server as call reports, and polls all spawned tasks to completion. The `utils.sync_repo` method uses the `api.Client` class, and therefore also waits for spawned tasks to complete. These two lines of code have no effect except to poll already-completed tasks, and the same is true in several other places.

I'll take care of this in a follow-up commit if these lines are left in. No big deal. But it may set your mind at ease to know that task polling is taken care of automatically.